### PR TITLE
Added seniority_zero core functions in new subpackage

### PR DESCRIPTION
**Pull request by obriente**
This is a first commit (13 files + 13 test files) for the code used in the seniority-zero / error mitigation experiment published in https://arxiv.org/abs/2210.10799 .

### CODE REVIEW

**Comment by obriente** [comment]
@mpharrigan any idea what's going on with the checks here?

**Comment by mpharrigan** [comment]
I think it's an anti-abuse measure for first-time contributors: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

I've approved the CI.

Please update the title of this PR to provide a self-contained one-line description of the change

File: recirq/seniority_zero/circuits_expt/core.py
Comment (mpharrigan) [inline recirq/seniority_zero/circuits_expt/core.py:36]: here and elsewhere: please ensure that the docstrings start with a one line summary optionally followed by a blank line and the rest of the docs.

This format is important for the docs system. The summary line must be one line. You can use 100 character line length

File: recirq/seniority_zero/circuits_expt/core.py
Comment (mpharrigan) [inline recirq/seniority_zero/circuits_expt/core.py:103]: return type annotation and/or documentation

File: recirq/seniority_zero/circuits_expt/core.py
Comment (mpharrigan) [inline recirq/seniority_zero/circuits_expt/core.py:116]: elsewhere you use a lowercase l `list`. Please be consistent. Note that it's not necessary to include the types in the docstring if you include them as type annotations in the function signature, which you do here

File: recirq/seniority_zero/circuits_expt/core.py
Comment (mpharrigan) [inline recirq/seniority_zero/circuits_expt/core.py:279]: docstring format

File: recirq/seniority_zero/circuits_expt/core.py
Comment (mpharrigan) [inline recirq/seniority_zero/circuits_expt/core.py:281]: please avoid `print` in library code

File: recirq/seniority_zero/circuits_expt/gates_this_experiment.py
Comment (mpharrigan) [inline recirq/seniority_zero/circuits_expt/gates_this_experiment.py:17]: this is really dirty lol. 

minimum bar: please use all caps variable name for `GATE_TYPE`, which is a global variable. Note that if you want to override it you must do so before importing this file. Please rename this file to something like `gate_type_import_delegation` or something weird and scary to denote that this is weird and scary.

A better way to structure this would be to set up a dataclass and two globals for the two gatesets


```python
@dataclass
class SeniorityZeroGateSet:
  cmnot: Gate
  cnot: Gate
  gsgate: Gate
  rzz: Gate
  swap_diag: Gate

SQRT_ISWAP_GATESET = SeniorityZeroGateSet(
  cmnot = cmnot_from_sqrt_swap,
 ..
)

CZ_GATESET = SeniorityZeroGateSet(
  cmnot = cmnot_from_cz,
  ...
)
```

Then to use it

```python
from .gate_sets import SeniorityZeroGateset, SQRT_ISWAP_GATESET, CZ_GATESET

def get_circuit_or_whatever(n: int, gs: SeniorityZeroGateset):
  for _ in range(n):
    circuit.append(gs.cmnot(i, j))

def driver_code():
   gateset = CZ_GATESET
   circuit = get_circuit(5, CZ_GATESET)
   ...
```

File: recirq/seniority_zero/data_processing/general.py
Comment (mpharrigan) [inline recirq/seniority_zero/data_processing/general.py:30]: This is too much. You can use like `*Note!*` or something to highlight importance but please use lowercase and no hashtags. This will probably mess up the markdown rendering too

File: recirq/seniority_zero/data_processing/general.py
Comment (mpharrigan) [inline recirq/seniority_zero/data_processing/general.py:38]: can we import these at the top level of the module?

File: recirq/seniority_zero/data_processing/storage.py
Comment (mpharrigan) [inline recirq/seniority_zero/data_processing/storage.py:25]: remove `#`s

**Review by mpharrigan — COMMENTED** [review]
left some initial comments.

Can you add return type annotations to (some of) the functions?

**Comment by mpharrigan** [comment]
Also, when you push new changes make sure to also ping me with a comment as pushing changes doesn't necessarily email me

File: recirq/seniority_zero/circuits_expt/gates_this_experiment.py
Comment (obriente) [reply recirq/seniority_zero/circuits_expt/gates_this_experiment.py]: Thanks for the tip, this is a far better idea than what I was doing. I'll push the change now and propagate it through with the latter imports.

File: recirq/seniority_zero/circuits_expt/core.py
Comment (obriente) [reply recirq/seniority_zero/circuits_expt/core.py:306]: Thanks for the catch!

File: recirq/seniority_zero/circuits_expt/core.py
Comment (obriente) [reply recirq/seniority_zero/circuits_expt/core.py]: I think I've gotten all of these.

File: recirq/seniority_zero/circuits_expt/core.py
Comment (obriente) [reply recirq/seniority_zero/circuits_expt/core.py]: I think I've gotten this

File: recirq/seniority_zero/circuits_expt/core.py
Comment (obriente) [reply recirq/seniority_zero/circuits_expt/core.py]: Fair enough. I've changed the 'list' to 'List' everywhere as that's what is in the docstring.

File: recirq/seniority_zero/circuits_expt/core.py
Comment (obriente) [reply recirq/seniority_zero/circuits_expt/core.py]: Fixed

File: recirq/seniority_zero/data_processing/general.py
Comment (obriente) [reply recirq/seniority_zero/data_processing/general.py]: Fixed - I think this was copied out of cirq and changed a bit lol.

File: recirq/seniority_zero/data_processing/storage.py
Comment (obriente) [reply recirq/seniority_zero/data_processing/storage.py]: Done

**Comment by obriente** [comment]
@mpharrigan - sorry to be a pain, can you approve the workflow again?

**Comment by mpharrigan** [comment]
That's really annoying. I've never seen it behave this way before

**Comment by mpharrigan** [comment]
@fdmalone can you give this a review

**Comment by mpharrigan** [comment]
Looks like you need additional requirements. ReCirq provides a basic set of default dependencies in  https://github.com/quantumlib/ReCirq/blob/master/requirements.txt. 

Please add a file in your subfolder named `extra-requirements.txt`. You can follow e.g. https://github.com/quantumlib/ReCirq/blob/master/recirq/hfvqe/extra-requirements.txt for an example.

You will also need to add the directory name to the list of subpackages with extra requirements in `dev_tools/write_ci_requirements.py` here: https://github.com/quantumlib/ReCirq/blob/master/dev_tools/write-ci-requirements.py#L77-L79

I think that should be sufficient and please don't hesitate to ping me with any questions. I tried to make it so it won't wait for approval to run the CI checks going forward

**Comment by obriente** [comment]
@mpharrigan - I'm now getting a test failing in qaoa because scipy.optimize expected a 1D array input. The test passes on my machine, however. Looking at the code, it seems that the issue is you have a list of 2 1x1 ndarrays and this is throwing an error? I'm trying to avoid this with np.concatenate, let's see if that works.

File: recirq/seniority_zero/circuits_expt/core.py
Comment (fdmalone) [inline recirq/seniority_zero/circuits_expt/core.py:34]: What is this a List of?

File: recirq/seniority_zero/circuits_simplified/core.py
Comment (fdmalone) [inline recirq/seniority_zero/circuits_simplified/core.py:48]: What is brickwall?

File: recirq/seniority_zero/circuits_simplified/core.py
Comment (fdmalone) [inline recirq/seniority_zero/circuits_simplified/core.py:51]: List of qubits or a loop of qubits?

File: recirq/seniority_zero/circuits_expt/gates_test.py
Comment (fdmalone) [inline recirq/seniority_zero/circuits_expt/gates_test.py:32]: Do we need a blanline here? I guess test functions aren't rendered in any docs.

File: recirq/seniority_zero/circuits_simplified/gates.py
Comment (fdmalone) [inline recirq/seniority_zero/circuits_simplified/gates.py:26]: Add a blank line before Args.

File: recirq/seniority_zero/circuits_simplified/gates.py
Comment (fdmalone) [inline recirq/seniority_zero/circuits_simplified/gates.py:81]: Blank line.

File: recirq/seniority_zero/circuits_simplified/gates.py
Comment (fdmalone) [reply recirq/seniority_zero/circuits_simplified/gates.py:77]: Oh, it's plus and minus right?

File: recirq/seniority_zero/data_processing/general.py
Comment (fdmalone) [inline recirq/seniority_zero/data_processing/general.py:33]: Missing Args: in docstring.

File: recirq/seniority_zero/data_processing/general.py
Comment (fdmalone) [inline recirq/seniority_zero/data_processing/general.py:316]: Missing Args: and an an indent for the args.

File: recirq/seniority_zero/data_processing/general.py
Comment (fdmalone) [inline recirq/seniority_zero/data_processing/general.py:379]: Same as above. also arg names don't match function signature.

File: recirq/seniority_zero/hamiltonians/rg_hamiltonian_test.py
Comment (fdmalone) [inline recirq/seniority_zero/hamiltonians/rg_hamiltonian_test.py:31]: Google style docstrings.

File: recirq/seniority_zero/misc.py
Comment (fdmalone) [inline recirq/seniority_zero/misc.py:266]: Can ignore, but there's a mix of """ and ''' doctstring quotes here.

File: recirq/seniority_zero/circuits_simplified/core.py
Comment (obriente) [reply recirq/seniority_zero/circuits_simplified/core.py:51]: Probably 'loop' here is better as this is the connectivity we assume. I've made a comment about this.

File: recirq/seniority_zero/circuits_expt/gates_test.py
Comment (obriente) [reply recirq/seniority_zero/circuits_expt/gates_test.py:32]: Easy enough to add :)

File: recirq/seniority_zero/circuits_simplified/gates.py
Comment (obriente) [reply recirq/seniority_zero/circuits_simplified/gates.py:26]: Done

File: recirq/seniority_zero/circuits_simplified/gates.py
Comment (obriente) [reply recirq/seniority_zero/circuits_simplified/gates.py:81]: Lol, also the docstring is just plain wrong. Fixed.

File: recirq/seniority_zero/circuits_simplified/gates.py
Comment (obriente) [reply recirq/seniority_zero/circuits_simplified/gates.py:158]: Docstring is also wrong, fixed.

File: recirq/seniority_zero/data_processing/general.py
Comment (obriente) [reply recirq/seniority_zero/data_processing/general.py:33]: Fixed

File: recirq/seniority_zero/data_processing/general.py
Comment (obriente) [reply recirq/seniority_zero/data_processing/general.py:316]: Fixed

File: recirq/seniority_zero/data_processing/general.py
Comment (obriente) [reply recirq/seniority_zero/data_processing/general.py:379]: Fixed.

File: recirq/seniority_zero/hamiltonians/rg_hamiltonian_test.py
Comment (obriente) [reply recirq/seniority_zero/hamiltonians/rg_hamiltonian_test.py:31]: Fixed here and below

File: recirq/seniority_zero/data_processing/general.py
Comment (obriente) [reply recirq/seniority_zero/data_processing/general.py:405]: Fixed

File: recirq/seniority_zero/scheduling.py
Comment (obriente) [reply recirq/seniority_zero/scheduling.py:27]: Thanks for catching. This really just rotates the initial list and returns a view of whatever was given. Probably this is too trivial to warrant its own function, but it was added and used in some parts of the code and I can't be bothered removing it lol.

**Comment by fdmalone** [comment]
LGTM. I imagine the scipy issue is probably a version update thing. concatenating seems fine, but possibly it should be pulled out into a separate PR since it touches the qaoa experiment? @mpharrigan what do you think?

**Comment by mpharrigan** [comment]
Since it's a small change and I've reviewed it and it's already in this PR let's just sneak it in.

**Review by mpharrigan — APPROVED** [review]
lgtm

**Pull request merged**
